File: app/src/main/java/com/hwonchul/movie/domain/usecase/listing/GetMovieListUseCaseImpl.java
package com.hwonchul.movie.domain.usecase.listing;

import androidx.annotation.NonNull;

import com.hwonchul.movie.domain.model.Movie;
import com.hwonchul.movie.domain.model.MovieListType;
import com.hwonchul.movie.domain.repository.MovieRepository;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.rxjava3.core.Flowable;

public class GetMovieListUseCaseImpl implements GetMovieListUseCase {
    private final MovieRepository repository;

    @Inject
    public GetMovieListUseCaseImpl(@NonNull MovieRepository repository) {
        this.repository = repository;
    }

    @Override
    public Flowable<List<Movie>> invoke(MovieListType listType) {
        return repository.getAllMoviesByListType(listType);
    }
}

File: app/src/main/java/com/hwonchul/movie/domain/usecase/listing/GetMovieListUseCaseImpl.kt
package com.hwonchul.movie.domain.usecase.listing

import com.hwonchul.movie.domain.model.Movie
import com.hwonchul.movie.domain.model.MovieListType
import com.hwonchul.movie.domain.repository.MovieRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.map
import javax.inject.Inject

class GetMovieListUseCaseImpl @Inject constructor(
    private val repository: MovieRepository
) : GetMovieListUseCase {

    override suspend fun invoke(listType: MovieListType): Flow<Result<List<Movie>>> {
        return repository.getAllMoviesByListType(listType)
            .map { Result.success(it) }
            .catch { emit(Result.failure(it)) }
    }
}